### PR TITLE
extmod/machine_uart: Factor UART to common extmod code.

### PR DESCRIPTION
**Pull request by dpgeorge**

### CODE REVIEW

**Comment by robert-hh** [comment]
The change seems fine. You follow the same approach as with PWM. I did not test the changes yet, but expect no difference. One question about readchar() and writechar(): what is the semantic difference to read() and write() for just one character? Is there a different timeout?

**Comment by dpgeorge** [comment]
> The change seems fine.

Thanks for the review.  I also did not run any tests on hardware yet.

> One question about readchar() and writechar(): what is the semantic difference to read() and write() for just one character? Is there a different timeout?

I'm not sure... these readchar/writechar methods are legacy and are only there to provide backwards compatibility for those ports that had them in the first place.  They should not be implemented on other ports.

These refactorings are intended to have no change in functionality for any of the ports.  Later on we can clean things up and make ports more consistent.

**Comment by iabdalkader** [comment]
This is a very good cleanup, I can tell by the amount of code removed, but in particular it's good to see `pyb` name changed to `machine` in the `stm32` ports, so now something like `cywbt.c` is (closer?) to being 100% portable. However, I was wondering why not use weak functions instead of including `.c` files ? Including the `.c` forces ports to implement every function, or provide empty ones, but more importantly weak functions can simplify the `MICROPY_PY_MACHINE_ADC` config macros by providing defaults that raise exceptions when they are not implemented (which is I think better than the "has no attr/function" exception that would be raised anyway).

**Comment by dpgeorge** [comment]
> However, I was wondering why not use weak functions instead of including `.c` files ?

Including .c files was done to minimise code size.  You can see that all these refactoring PRs have almost zero change in code size, because the implementation functions provided by the ports can be inlined.

To get around this issue of code size, we could enable LTO on all ports... but I'm not sure if we are ready to do that just yet.

At the moment a port can configure which things it implements with macro options, so there's no real need for weak functions (if a port doesn't implement the feature it just doesn't turn it on).

> providing defaults that raise exceptions when they are not implemented (which is I think better than the "has no attr/function" exception that would be raised anyway).

It would increase code size by raising specific exceptions for each method that's not implemented.  And the current exception for a missing method is actually pretty good, eg: it would currently raise `AttributeError: type object 'UART' has no attribute 'writechar'`.  I don't think a custom message could be much more specific than that.

**Comment by iabdalkader** [comment]
Those are valid points, except for the fine-grained machine config macros, reducing the config macros would be nice, but it's fine otherwise.

**Comment by dpgeorge** [comment]
> reducing the config macros would be nice

I agree, the config options should be eventually reduced.  But for now the aim is to refactor without changing behaviour, and then we'll arrive at a point where we can easily see what all the differences are among the ports, and then fix it.

File: ports/esp8266/machine_uart.c [hunk 1]
```diff
@@ -214,47 +211,30 @@ STATIC mp_obj_t pyb_uart_make_new(const mp_obj_type_t *type, size_t n_args, size
...
-    { MP_ROM_QSTR(MP_QSTR_read), MP_ROM_PTR(&mp_stream_read_obj) },
-    { MP_ROM_QSTR(MP_QSTR_readline), MP_ROM_PTR(&mp_stream_unbuffered_readline_obj) },
-    { MP_ROM_QSTR(MP_QSTR_readinto), MP_ROM_PTR(&mp_stream_readinto_obj) },
-    { MP_ROM_QSTR(MP_QSTR_write), MP_ROM_PTR(&mp_stream_write_obj) },
-};
-
-STATIC MP_DEFINE_CONST_DICT(pyb_uart_locals_dict, pyb_uart_locals_dict_table);
+STATIC void mp_machine_uart_sendbreak(machine_uart_obj_t *self) {
+    (void)self;
+    // TODO
```
Comment (projectgus) [inline ports/esp8266/machine_uart.c:233]: If I follow correctly, the previous UART object didn't have this function and now it's a no-op? Suggest `mp_raise_NotImplementedError(NULL); // TODO` instead.

File: ports/nrf/modules/machine/uart.c [hunk 2]
```diff
@@ -267,78 +264,41 @@ STATIC mp_obj_t machine_hard_uart_make_new(const mp_obj_type_t *type, size_t n_a
...
-/*
-    { MP_ROM_QSTR(MP_QSTR_RTS), MP_ROM_INT(UART_HWCONTROL_RTS) },
-    { MP_ROM_QSTR(MP_QSTR_CTS), MP_ROM_INT(UART_HWCONTROL_CTS) },
-*/
-};
 
-STATIC MP_DEFINE_CONST_DICT(machine_hard_uart_locals_dict, machine_hard_uart_locals_dict_table);
+// uart.sendbreak()
+STATIC void mp_machine_uart_sendbreak(machine_uart_obj_t *self) {
+    (void)self;
```
Comment (projectgus) [inline ports/nrf/modules/machine/uart.c:297]: Same suggestion about `NotImplementedError` for this one.

File: ports/stm32/mpconfigport.h [hunk 3]
```diff
@@ -123,6 +123,10 @@
 #define MICROPY_PY_MACHINE_SOFTSPI  (1)
 #define MICROPY_PY_MACHINE_TIMER    (1)
 #endif
+#define MICROPY_PY_MACHINE_UART     (1)
+#define MICROPY_PY_MACHINE_UART_INCLUDEFILE "ports/stm32/machine_uart.c"
+#define MICROPY_PY_MACHINE_UART_IRQ (1)
+#define MICROPY_PY_MACHINE_UART_READCHAR_WRITECHAR (1)
```
Comment (projectgus) [inline ports/stm32/mpconfigport.h:129]: Suggest adding a block somewhere (probably `extmod/machine_uart.h`) that does `#ifndef MICROPY_PY_MACHINE_UART_READCHAR_WRITECHAR; #define MICROPY_PY_MACHINE_UART_READCHAR_WRITECHAR (0)` for all of these, and includes any relevant comments on what they are for.

(This is another suggestion that doesn't change anything at compile time, but it gives the implementer somewhere to quickly see all of the macros they may or may not want to tweak in their port.)

File: ports/zephyr/machine_uart.c [hunk 4]
```diff
@@ -84,20 +86,31 @@ STATIC mp_obj_t machine_uart_make_new(const mp_obj_type_t *type, size_t n_args,
...
-};
-STATIC MP_DEFINE_CONST_DICT(machine_uart_locals_dict, machine_uart_locals_dict_table);
+STATIC void mp_machine_uart_deinit(machine_uart_obj_t *self) {
+    (void)self;
+}
 
-STATIC mp_uint_t machine_uart_read(mp_obj_t self_in, void *buf_in, mp_uint_t size, int *errcode) {
+STATIC mp_int_t mp_machine_uart_any(machine_uart_obj_t *self) {
+    (void)self;
+    return 0; // TODO
```
Comment (projectgus) [inline ports/zephyr/machine_uart.c:100]: Suggest this one and `mp_machine_uart_sendbreak()` raise `NotImplementedError` as well. Much better to error out than silently appear to work but do nothing.

File: ports/zephyr/machine_uart.c [hunk 5]
```diff
@@ -84,20 +86,31 @@ STATIC mp_obj_t machine_uart_make_new(const mp_obj_type_t *type, size_t n_args,
...
+}
+
+STATIC bool mp_machine_uart_txdone(machine_uart_obj_t *self) {
+    (void)self;
+    return true;
```
Comment (projectgus) [inline ports/zephyr/machine_uart.c:105]: Suggest this either raises `NotImplementedError` or has a comment about why it's always true (does Zephyr TX never return before done?)

**Review by projectgus — COMMENTED** [review]
I wanted to take a look at this one after #12646 as I was curious how the appoach scales to more complex peripherals. It scales well, I think!

Apologies if any of these comments are pre-emptive on things you were going to get back to in this PR, already.

File: ports/esp8266/machine_uart.c (hunk 1 above)
Comment (projectgus) [reply ports/esp8266/machine_uart.c:233]: (After going through all the ports, maybe there's enough ports without `sendbreak` that it's worth making it a config macro flag as well?)

Comment (dpgeorge) [reply ports/esp8266/machine_uart.c]: OK, I made sendbreak configurable via a macro.

File: ports/stm32/mpconfigport.h (hunk 3 above)
Comment (dpgeorge) [reply ports/stm32/mpconfigport.h:134]: Added this to `extmod/modmachine.h`.

File: ports/zephyr/machine_uart.c (hunk 4 above)
Comment (dpgeorge) [reply ports/zephyr/machine_uart.c]: This now raises.

File: ports/zephyr/machine_uart.c (hunk 5 above)
Comment (dpgeorge) [reply ports/zephyr/machine_uart.c]: This now raises.

**Review by projectgus — COMMENTED** [review]
image
You love to see it. :grinning:

**Pull request merged**